src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureAnalyzer.cs

```diff
@@ -0,0 +1,304 @@
+                }
+
+                context.ReportDiagnostic(operation.CreateDiagnostic(GeneralPreviewFeatureAttributeRule, symbol.Name));
+            }
+        }
+
+        private static bool OperationUsesPreviewFeatures(OperationAnalysisContext context, ConcurrentDictionary<ISymbol, bool> requiresPreviewFeaturesSymbols, [NotNullWhen(true)] out ISymbol? symbol)
+        {
+            IOperation operation = context.Operation;
+            if (operation is IInvocationOperation || operation is IPropertyReferenceOperation)
+            {
+                ISymbol containingSymbol = context.ContainingSymbol;
+                if (OperationUsesPreviewFeatures(containingSymbol, requiresPreviewFeaturesSymbols) || OperationUsesPreviewFeatures(containingSymbol.ContainingSymbol, requiresPreviewFeaturesSymbols))
+                {
+                    symbol = null;
+                    return false;
+                }
+            }
+
+            symbol = GetOperationSymbol(operation);
+            return symbol != null && OperationUsesPreviewFeatures(symbol, requiresPreviewFeaturesSymbols);
+        }
+
+        private static ISymbol? GetOperationSymbol(IOperation operation)
+            => operation switch
+            {
+                IInvocationOperation iOperation => iOperation.TargetMethod,
+                IObjectCreationOperation cOperation => cOperation.Constructor,
+                IPropertyReferenceOperation pOperation => pOperation.Property,
+                IFieldReferenceOperation fOperation => fOperation.Field,
+                IDelegateCreationOperation dOperation => dOperation.Type,
+                IEventReferenceOperation eOperation => eOperation.Member,
+                IUnaryOperation uOperation => uOperation.OperatorMethod,
+                IBinaryOperation bOperation => bOperation.OperatorMethod,
+                IArrayCreationOperation arrayCreationOperation => (arrayCreationOperation.Type as IArrayTypeSymbol)?.ElementType,
+                ICatchClauseOperation catchClauseOperation => catchClauseOperation.ExceptionType,
+                _ => null,
+            };
+
+        private static bool OperationUsesPreviewFeatures(ISymbol symbol, ConcurrentDictionary<ISymbol, bool> requiresPreviewFeaturesSymbols)
+        {
+            if (ContainingTypeIsPreview(symbol, requiresPreviewFeaturesSymbols))
+            {
+                return true;
+            }
+
+            return ProcessPreviewAttribute(symbol, requiresPreviewFeaturesSymbols);
+        }
+
+        private static bool ProcessPreviewAttribute(ISymbol symbol, ConcurrentDictionary<ISymbol, bool> requiresPreviewFeaturesSymbols)
+        {
+            if (!requiresPreviewFeaturesSymbols.TryGetValue(symbol, out bool existing))
+            {
+                bool ret = false;
+                // TODO: Use symbol.HasAttribute(RequiresPreviewFeaturesAttribute) once we consume the latest sdk
```

Comment:
Can you file a tracking bug with all the TODOs and assigned it to yourself?

Reply:
https://github.com/dotnet/roslyn-analyzers/issues/5227